examples/src/main/java/kafka/examples/ExactlyOnceMessageProcessor.java

```diff
@@ -81,111 +88,115 @@ public ExactlyOnceMessageProcessor(final String inputTopic,
-                // if the producer has hit a fatal error.
-                producer.abortTransaction();
-
-                // The consumer fetch position needs to be restored to the committed offset
-                // before the transaction started.
-                resetToLastCommittedPositions(consumer);
-            }
+    @Override
+    public void onPartitionsLost(Collection<TopicPartition> partitions) {
+        Utils.printOut("Lost partitions: %s", partitions);
+    }
 
-            messageRemaining.set(messagesRemaining(consumer));
-            printWithTxnId("Message remaining: " + messageRemaining);
+    public void shutdown() {
+        if (!closed) {
+            closed = true;
+            latch.countDown();
         }
-
-        printWithTxnId("Finished processing " + messageProcessed + " records");
-        latch.countDown();
     }
 
-    private Map<TopicPartition, OffsetAndMetadata> consumerOffsets() {
+    private Map<TopicPartition, OffsetAndMetadata> getOffsetsToCommit(KafkaConsumer<Integer, String> consumer) {
         Map<TopicPartition, OffsetAndMetadata> offsets = new HashMap<>();
         for (TopicPartition topicPartition : consumer.assignment()) {
             offsets.put(topicPartition, new OffsetAndMetadata(consumer.position(topicPartition), null));
         }
         return offsets;
     }
 
-    private void printWithTxnId(final String message) {
-        System.out.println(transactionalId + ": " + message);
-    }
-
-    private ProducerRecord<Integer, String> transform(final ConsumerRecord<Integer, String> record) {
-        printWithTxnId("Transformed record (" + record.key() + "," + record.value() + ")");
-        return new ProducerRecord<>(outputTopic, record.key() / 2, "Transformed_" + record.value());
-    }
-
-    private long messagesRemaining(final KafkaConsumer<Integer, String> consumer) {
+    private long getRemainingRecords(KafkaConsumer<Integer, String> consumer) {
         final Map<TopicPartition, Long> fullEndOffsets = consumer.endOffsets(new ArrayList<>(consumer.assignment()));
-        // If we couldn't detect any end offset, that means we are still not able to fetch offsets.
+        // if we can't detect any end offset, that means we are still not able to fetch offsets
         if (fullEndOffsets.isEmpty()) {
             return Long.MAX_VALUE;
         }
-
         return consumer.assignment().stream().mapToLong(partition -> {
             long currentPosition = consumer.position(partition);
-            printWithTxnId("Processing partition " + partition + " with full offsets " + fullEndOffsets);
             if (fullEndOffsets.containsKey(partition)) {
                 return fullEndOffsets.get(partition) - currentPosition;
+            } else {
+                return 0;
             }
-            return 0;
         }).sum();
     }
-
-    private static void resetToLastCommittedPositions(KafkaConsumer<Integer, String> consumer) {
```

Comment:
Sorry, I have a question. Why remove this? We don't need to reset the offset manually, right?

Reply:
> We don't need to reset the offset manually, right?

This is my understanding. Here we are subscribing, and the Consumer Javadoc says:

"On each poll, consumer will try to use the last consumed offset as the starting offset and fetch sequentially. The last consumed offset can be manually set through seek(TopicPartition, long) or automatically set as the last committed offset for the subscribed list of partitions."

Reply:
I did some tests to set `max.poll.records` to 10. I tried to abort the transaction randomly, but the consumer didn't poll aborted messages automatically; I guess this is not match exactly once semantic; maybe we also need to seek topics to LSO manually while aborting transactions, WDYT?

Reply:
I did a similar experiment by alternating commit and abort 1 record at a time, and you are correct. That said, if the transaction is always aborted (e.g. a poison pill) the example would not make any progress. Instead, the current logic moves on with the next batch. In a real world application you could send failed records to a DLT (dead letter topic) for further processing. Here we can simply log them and add a comment, WDYT?

Reply:
> In a real world application you could send failed records to a DLT (dead letter topic) for further processing.

Yes, in commons, we can do some retries (such as five times); if they always fail to process, we can send them to DLT.

> Here we can simply log them and/or add a comment.

If we can add a comment, it's better. Otherwise, users may be confused.

Reply:
Opened https://github.com/apache/kafka/pull/15561, let's see what others think.